src/appleseed/renderer/modeling/postprocessingstage/vignettepostprocessingstage.cpp

```diff
@@ -0,0 +1,198 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "vignettepostprocessingstage.h"
+
+// appleseed.renderer headers.
+#include "renderer/modeling/frame/frame.h"
+#include "renderer/modeling/postprocessingstage/postprocessingstage.h"
+
+// appleseed.foundation headers.
+#include "foundation/containers/dictionary.h"
+#include "foundation/image/canvasproperties.h"
+#include "foundation/image/color.h"
+#include "foundation/image/image.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+
+using namespace foundation;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // Vignette post-processing stage.
+    //
+
+    const char* Model = "vignette_post_processing_stage";
+
+    const float DefaultIntensity = 0.5f;
+    const float MinIntensity = 0.0f;
+    const float MaxIntensity = 1.0f;
+
+    class VignettePostProcessingStage
+        : public PostProcessingStage
+    {
+      public:
+        VignettePostProcessingStage(
+            const char*             name,
+            const ParamArray&       params)
+          : PostProcessingStage(name, params)
+        {
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        const char* get_model() const override {
+            return Model;
+        }
+
+        bool on_frame_begin(
+            const Project&          project,
+            const BaseGroup*        parent,
+            OnFrameBeginRecorder&   recorder,
+            IAbortSwitch*           abort_switch) override
+        {
+            const OnFrameBeginMessageContext context("post-processing stage", this);
+
+            m_intensity = m_params.get_optional("intensity", DefaultIntensity, context);
+
+            return true;
+        }
+
+        void execute(Frame& frame) const override
+        {
+            const CanvasProperties& props = frame.image().properties();
+            const Vector2f resolution(static_cast<float>(props.m_canvas_width), static_cast<float>(props.m_canvas_height));
+            const Vector2f aspect(resolution.x / resolution.y, 1);
+
+            Image& image = frame.image();
+
+            for (size_t y = 0; y < resolution.y; ++y)
```

Comment:
You're mixing up float and integer here. Use `props.m_canvas_height` for the upper bound, same for `x`.